core/camel-support/src/main/java/org/apache/camel/support/task/BackgroundTask.java

```diff
@@ -167,12 +147,14 @@ private boolean waitForTaskCompletion(CountDownLatch latch, Future<?> task) {
         } finally {
             elapsed = budget.elapsed();
         }
-
-        return completed;
     }
 
     @Override
     public Duration elapsed() {
         return elapsed;
     }
+
+    private static class ExecutionResult {
+        private boolean completed;
```

Comment:
@orpiske This makes me wonder if `BackgroundTask` is stateful or stateless. I mean can run several times the same task? because if not, I guess that the `CountDownLatch` and this field could be moved directly as fields of `BackgroundTask` otherwise maybe we could add the `CountDownLatch` to that class and rename it to something like `ExecutionState`

Reply:
Yes, it's stateful because of `budget` and `elapsed`.

I think your suggestion is +/- in line with what I commented on the Zulip chat: to investigate moving the latch into the budget and consider it as part of that. Granted, I did not comment on the elapsed, but I think it would eventually become a natural move as well. 

IMHO, making (all) the tasks stateless - keeping the state on the budget itself - could potentially be a better approach for the future (maybe as part of Camel 4?), but I think the current approach is fine.